load_confounds/confounds.py

```diff
@@ -217,24 +216,59 @@ def _confounds_to_df(image_file, flag_acompcor, flag_full_aroma):
     return confounds_raw, confounds_json
 
 
-def _confounds_to_ndarray(confounds, demean):
-    """Convert confounds from a pandas dataframe to a numpy array."""
-    # Convert from DataFrame to numpy ndarray
-    labels = confounds.columns
-    confounds = confounds.values
+def _get_outlier_cols(confounds_columns):
+    """Get outlier regressor column names."""
+    outlier_cols = {
+        col
+        for col in confounds_columns
+        if "motion_outlier" in col or "non_steady_state" in col
+    }
+    confounds_col = set(confounds_columns) - outlier_cols
+    return outlier_cols, confounds_col
 
-    # Derivatives have NaN on the first row
-    # Replace them by estimates at second time point,
-    # otherwise nilearn will crash.
-    if confounds.size != 0:  # ica_aroma = "full" generate empty output
-        mask_nan = np.isnan(confounds[0, :])
-        confounds[0, mask_nan] = confounds[1, mask_nan]
 
-        # Optionally demean confounds
-        if demean:
-            confounds = scale(confounds, axis=0, with_std=False)
+def _extract_outlier_regressors(confounds):
+    """Separate confounds and outlier regressors."""
+    outlier_cols, confounds_col = _get_outlier_cols(confounds.columns)
+    outliers = confounds[outlier_cols] if outlier_cols else pd.DataFrame()
+    confounds = confounds[confounds_col]
+    sample_mask = _outlier_to_sample_mask(outliers)
+    return sample_mask, confounds, outliers
 
-    return confounds, labels
+
+def _outlier_to_sample_mask(outlier_flag):
+    """Generate sample mask from outlier regressors."""
+    if outlier_flag.size == 0:  # Do not supply sample mask
+        return None  # consistency with nilearn sample_mask
+    outlier_flag = outlier_flag.sum(axis=1).values
+    return np.where(outlier_flag == 0)[0].tolist()
+
+
+def _confounds_to_ndarray(confounds, demean, scrub_mask):
+    """Convert confounds from a pandas dataframe to a numpy array."""
+    sample_mask, confounds, outliers = _extract_outlier_regressors(confounds)
+    if confounds.size != 0:  # ica_aroma = "full" generate empty output
+        # Derivatives have NaN on the first row
+        # Replace them by estimates at second time point,
+        # otherwise nilearn will crash.
+        mask_nan = np.isnan(confounds.values[0, :])
+        confounds.iloc[0, mask_nan] = confounds.iloc[1, mask_nan]
+        if demean:
+            confounds = _demean_confounds(confounds, outliers, scrub_mask)
+    labels = list(confounds.columns)
+    confounds = confounds.values
+    return sample_mask, confounds, labels
+
+
+def _demean_confounds(confounds, outliers, scrub_mask):
+    """Demean the confounds."""
+    confound_cols = confounds.columns
+    confounds = scale(confounds, axis=0, with_std=False)
+    confounds = pd.DataFrame(confounds, columns=confound_cols)
+    # using despike strategy (currently not supported)
```

Comment:
should we remove this bit of code as it's not supported?

Reply:
Yeah let's just remove them